### ISSUE TITLE
about precipitation

### ISSUE THREAD

**Issue by Passion28-y** [LLNL/ParaDiS#7]
I can use the precipitation file in the bin folder to generate a precipitation file, but I cannot call this file in the ctrl file. I did not find any parameters to specify this file. In addition, when generating precipitation, can't I specify the size and position of the precipitation? I have been thinking about this question for a long time. I hope you can answer my confusion. Thank you very much.

**Comment by nrbertin** [comment]
In ParaDiS precipitates are represented as Eshelby inclusions. To activate this feature, first make sure the code is compiled with option `-DESHELBY` (line 336 of `makefile.setup`). Then you can take a look at example `tests/onescrew.ctrl`. In this control file you can see that the precipitate/inclusion file is specified with option `inclusionFile`, e.g. `inclusionFile = "tests/onescrew.dat"`. In this example, the precipitates configuration is defined in file `tests/onescrew.dat`, where each line specifies a precipitate using the following format:
```
#
# Data for an precipitate consists of the following items separated with white space
#
# Precipitate ID     : 1 integer identifying the precipitate regardless of its position in the simulation or the domain encompassing it. (Should be sequential values)
# Position           : 3 values specifying the XYZ coordinates of the center of the precipitate
# Semi-principal axes: 3 values defining the three semi-principal axis of the ellipsoidal precipitate in units of b
# Rotation matrix    : 2 vectors defining the rotation of the ellipsoidal precipitates
# Strain field       : 6 components of the strain field for the particleStrain field is a symmetric matrix so only six components are specified
#
```
- If you want to place precipitates at very specific locations then you can generate the precipitate file manually following the format above.
- For randomly positioned spherical inclusions you can use tool `bin/inclusiongen`
- For randomly positioned more specific types of inclusions (e.g. ellispoid T or ThetaPrime precipitates) you can use tool `bin/precipitates`

**Comment by Passion28-y** [comment]
@nrbertin I understand. Words are powerless to express my gratitude.Thank you!

**Comment by Passion28-y** [comment]
@nrbertin Hello, in ParaDiS, I added the precipitation phase simulation. However, I cannot observe the precipitation phase using the visualization software visit. I wonder if there is a problem with my operation. Can you help me solve it? Thank you

**Comment by nrbertin** [comment]
@Passion28-y Unfortunately the precipitates configuration is not written out in the output files, so it won't display when visualizing the dislocation configuration.

**Comment by Passion28-y** [comment]
@nrbertin I see, thank you for your answer.

**Comment by Passion28-y** [comment]
@nrbertin Hello, developer. I found that the initial model generated by the paradisgen tool is too random. Is there any way to generate a model that meets my needs according to my own ideas? For example, I want to create a dislocation line that only moves along a certain direction of a certain surface, and there is no dislocation line in other places. I read the OpenDiS document, but I don’t understand the method described in the document. Is there any other way to generate the initial dislocation configuration? Looking forward to your reply.

**Comment by nrbertin** [comment]
@Passion28-y There is no tool provided with ParaDiS to create dislocation configurations besides `paradisgen`. However, it should be easy to create arbitrary configurations using [OpenDiS](https://github.com/OpenDiS/OpenDiS). I have added a new [initial configuration](https://opendis.github.io/OpenDiS/tutorials/initial_configuration.html) tutorial section to provide more details about the different ways to do so. For setting a dislocation that only moves along a given direction on a given plane, the [Rotated crystal frames](https://opendis.github.io/OpenDiS/tutorials/initial_configuration.html#rotated-crystal-frames) subsection may be useful. I hope this helps.

**Comment by Passion28-y** [comment]
@nrbertin Ok, thank you very much.

**Comment by Passion28-y** [comment]
@nrbertin Dear developer
Can precipitated phases define strength?

**Comment by nrbertin** [comment]
@Passion28-y Not sure if this is what you are asking about, but the presence of precipitates will certainly have an effect on the predicted strength.

**Comment by Passion28-y** [comment]
@nrbertin I am now simulating the interaction between dislocations and precipitates. I found that dislocations can always cut precipitates, but there is also a way to bypass precipitates, which is the orowan bypass mechanism. So I want to know if it is possible to define the strength of precipitates? For example, harder precipitates or softer precipitates.

**Comment by nrbertin** [comment]
@Passion28-y You can use control parameters `shearModulus2` and `pois2` to define the elastic parameters associated with the inclusions. To precisely account for elastic interactions (forces) between dislocations and precipitates, one needs to compile the code with option -DESHELBYFORCE (line 342 of `makefile.setup`).

**Comment by Passion28-y** [comment]
@nrbertin Thank you very much.I'll try it.

**Comment by Passion28-y** [comment]
@nrbertin Hello developer, I am now using ParaDiS to simulate the effect of precipitation on CRSS in nickel-based high-temperature alloys. ParaDiS does not have other strengthening and can only consider precipitation strengthening. 
The first question , can I take other strengthening into account by reducing the mobility in the ctrl file? 
The second question , in post-processing, I don’t know what software to use to see the process demonstration of precipitation phase and dislocation line. Currently, I can only use vislt to see the demonstration process of dislocation line, but I can’t see the precipitation phase. 
The third question , which parameters in the ctrl file have a greater impact on CRSS in the simulation? Because I have now simulated CRSS at different radii, but did not get the peak value, I don’t know if you have simulated nickel-based high-temperature alloys. What is the elastic modulus of the precipitation phase set in ctrl? 
I’m sorry to bother you, but this has been a problem that has been bothering me. I hope to get your reply, thank you.

**Comment by Passion28-y** [comment]
@nrbertin Another question, what is the difference between the dat generated by inclusiongen and precipitate? Is the precipitate phase treated as a rigid body? If it is an inclusion, can soft and hard inclusions be considered?

**Comment by nrbertin** [comment]
@Passion28-y 
1. Sure, reducing the mobility (e.g. to account for strengthening due to the presence of solutes in an implicit way) is an approach that has been taken in the past.

2. Utility [inputs/inclusionFile2ellipsoidVTK.py](https://github.com/LLNL/ParaDiS/blob/main/inputs/inclusionFile2ellipsoidVTK.py) should allow you to convert your inclusion.dat file into a vtk file that can be loaded into VisIt and superimposed onto the dislocation lines. You will need a vtk-enabled python interpreter to run it. The simplest is to use the one provided in Paraview. To run the script, simply give it the path to the python interpreter, e.g.:
`/path/to/ParaView.app/Contents/bin/pvpython inclusionFile2ellipsoidVTK.py -i inclusion.dat -o inclusion.vtk`
which will create file `inclusion.vtk` that you can then load in VisIt.

3. For a given strain rate, I think the mobility should have the greatest impact on strength. The shear modulus should play a role as well. The elastic modulus of precipitates is defined with variable `shearModulus2` in the ctrl file.

4. The only difference between between `inclusiongen` and `precipitates` is that `inclusiongen` produces generic spherical inclusions, while `precipitates` is meant to generate inclusions corresponding to specific types of precipitates (e.g. ellispoid T or ThetaPrime precipitates), i.e. having specific ellipsoidal aspect ratios oriented along specific crystal directions.

**Comment by Passion28-y** [comment]
@nrbertin Thank you very much for your reply, I will put it into practice first.

**Comment by Passion28-y** [comment]
@nrbertin The output files currently supported do not support Paraview. How can I view the interaction between dislocation lines and precipitated phases in Paraview? I can see the entire simulation process. I hope to get your reply. Thank you.

**Comment by nrbertin** [comment]
@Passion28-y Have you tried to use script `inclusionFile2ellipsoidVTK.py` I mentioned in my previous message?

**Comment by Passion28-y** [comment]
@nrbertin Yes, I have obtained the vtk file of the precipitated phase, so can I use vislt to view the simulation process? Is vislt the best solution?

**Comment by nrbertin** [comment]
@Passion28-y Yes, you should be able to load both the .data dislocation files and the .vtk precipitate file into VisIt.

If you want to use Paraview, then you'd have to convert the .data files into .vtk files first. This can be done using OpenDiS, following what is described in this [thread](https://github.com/OpenDiS/OpenDiS/issues/4).

**Comment by Passion28-y** [comment]
@nrbertin I will use OpenDiS to convert data to vtk. I have a question, can I only convert one data file at a time? Can I not convert in batches?

**Comment by nrbertin** [comment]
@Passion28-y Simply add a loop over all .data files to convert them in batch, e.g.:

```
import os, glob
import pyexadis
from pyexadis_utils import read_paradis, write_vtk
pyexadis.initialize()

datafile_path = '/path/to/paradis/simulation/restart'
for f in glob.glob(datafile_path+'/*.data'):
    N = read_paradis(f)
    write_vtk(N, os.path.splitext(f)[0]+'.vtk')
```

**Comment by Passion28-y** [comment]
@nrbertin OK, thank you very much for your reply, I will try it

**Comment by Passion28-y** [comment]
@nrbertin Excuse me again, why do the dislocation lines still cut through even though I set the precipitation phase strength to be very high? Is it because ParaDiS cannot implement the Orowan mechanism? Do I need to modify the MobilityLaw file to implement it?

**Comment by nrbertin** [comment]
@Passion28-y ParaDiS can certainly reproduce the Orowan mechanism. Are eigenstrain values for the precipitates specified in your precipitate .dat file? If not, the force exerted by precipitates on the dislocations will evaluate to 0.

**Comment by Passion28-y** [comment]
@nrbertin This is my dat file, I guess it should be unspecified

<img width="929" alt="Image" src="https://github.com/user-attachments/assets/6cc245db-6d39-4d95-b248-25848c3cbd17" />

**Comment by Passion28-y** [comment]
@nrbertin There is another question that has puzzled me for a long time, I have read a lot of literature that mentions Anti-phase boundary energy, but I don't see it mentioned in the ParaDiS manual, can this be set in ParaDiS?

**Comment by nrbertin** [comment]
@Passion28-y The six values of the eigenstrain are indeed set to zero in your .dat file (six last values). You need to specify them in the form
```
e e e 0.0 0.0 0.0
```
where e is the eigenstrain associated with a dilatational misfit.

No, there is no anti-phase boundary energy implemented in ParaDiS.

**Comment by Passion28-y** [comment]
@nrbertin OK, I got it, thank you

**Comment by Passion28-y** [comment]
@nrbertin Hello,When the radius is large and there are only dozens of precipitates, the plastic strain decreases after relaxation, and normal results can be simulated normally. However, when the radius decreases and the precipitates increase, the plastic strain does not change during relaxation, and reasonable results cannot be simulated. However, the volume fraction of the precipitates is consistent. Is this normal? Is it not possible to set too many precipitates?

**Comment by Passion28-y** [comment]
@nrbertin I seem to have solved this problem. It seems that it was due to the problem of randomly generating the position of the precipitated phase. I regenerated the dat file and changed the position of the precipitated phase, which solved the problem.

**Comment by Passion28-y** [comment]
> [@Passion28-y](https://github.com/Passion28-y) The six values of the eigenstrain are indeed set to zero in your .dat file (six last values). You need to specify them in the form
> 
> ```
> e e e 0.0 0.0 0.0
> ```
> 
> where e is the eigenstrain associated with a dilatational misfit.
> 
> No, there is no anti-phase boundary energy implemented in ParaDiS.

@nrbertin Hello, I encountered a problem when using ParaDiS, and I hope you can help me. Because the lattice constant of the precipitate phase in a certain direction is equal to the lattice constant of the matrix. The lattice mismatch of the precipitate phase can be defined in the dat file. How can I ensure that the lattice constant of the precipitate phase is equal to that of the matrix? Does it mean that the lattice constant of the precipitate phase is equal to that of the matrix if the strain field is directly set to 0 in the dat file?

**Comment by nrbertin** [comment]
You are correct, if there is no lattice mismatch, then the eigenstrain field equals 0. The approximation used in ParaDiS is that the precipitate produces a static stress field that is only proportional to its eigenstrain and independent of the dislocation microstructure. It means that if the eigenstrain is 0, then the precipitate will not exert any force on the dislocations.

Obviously this is an approximation, as in reality, the dislocations stress fields interact with the inclusions as well, which produces an additional elastic interaction force. For instance, even if there is no lattice mismatch, you can still get a dislocation/precipitate interaction force due to an elastic mismatch. But this will not be captured by the inclusion treatment currently implemented in ParaDiS.

**Comment by Passion28-y** [comment]
> You are correct, if there is no lattice mismatch, then the eigenstrain field equals 0. The approximation used in ParaDiS is that the precipitate produces a static stress field that is only proportional to its eigenstrain and independent of the dislocation microstructure. It means that if the eigenstrain is 0, then the precipitate will not exert any force on the dislocations.
> 
> Obviously this is an approximation, as in reality, the dislocations stress fields interact with the inclusions as well, which produces an additional elastic interaction force. For instance, even if there is no lattice mismatch, you can still get a dislocation/precipitate interaction force due to an elastic mismatch. But this will not be captured by the inclusion treatment currently implemented in ParaDiS.
> 

I understand. Thanks.

**Comment by Passion28-y** [comment]
@nrbertin nrbertin Hello, I am now using Eshably inclusions to consider coherent strengthening. Since the size of the precipitate is fixed, for example (10nm-50nm). When I use Lbox of 1000nm or 2000nm, the volume fraction is kept at about 0.12. If 10nm is used, the number of precipitates will reach 10 million, which is a considerable number.
My question is, if coherent strengthening is considered, can Lbox be smaller? For example, Lbox=100nm or Lbox=200nm. Can you give me some suggestions?
Question 2: When I use useLabFrame in the ctrl file, in order to keep the orientation of the precipitate consistent with that of the matrix, does the Rotation matrix in the dat file also need to be changed accordingly?

**Comment by nrbertin** [comment]
1. It depends what you are examining. If you're looking at unit mechanisms, e.g. one dislocation going through a forest of inclusions, then using a smaller box is probably fine. If you're looking at strain-hardening simulations, then you need to ensure that the volume is large enough to be statistically representative.
2. No, there is no need to specify the rotation matrix in inclusion dat file. The crystal orientation in the inclusion is always the same as that in the matrix. The rotation matrix in the dat file is only to orient the principal axes of ellipsoidal inclusions.

**Comment by Passion28-y** [comment]
Oh, I think I should ask a more specific question. Assume that in the default coordinate system, the direction of the principal axis of the ellipsoidal inclusion is consistent with the z-axis direction of the matrix. Now use the user coordinate system. Then will the principal axis of the inclusion also change? Does it mean that the rotation matrix needs to be redefined in dat?

**Comment by nrbertin** [comment]
No, I'm pretty sure that, as for the .data file, directions in the inclusion .dat files are always expressed in the global frame. That is, if your inclusion is elongated along the z-direction in the .dat file, then it will always remain elongated along z even if you specify some user coordinate system in the ctrl file. If you want the inclusion to rotate along with your user coordinate system, you'll have to manually rotate the inclusion orientation matrix in the .dat file.

**Comment by Passion28-y** [comment]
I understand, thank you

**Comment by Passion28-y** [comment]
In the user coordinate system, the crystal orientation of the matrix will also change, right?

**Comment by nrbertin** [comment]
Basically, there is a unique crystal orientation for the whole simulation box, i.e. it is the same inside and outside of the inclusions. This unique orientation can be changed using the user coordinate system.

**Comment by Passion28-y** [comment]
I understand, thank you
